[DM-28897] Initial version (#1)
lsst-dm/dmtn-182: index.rst

:tocdepth: 1

.. sectnum::

Abstract
========

> Given the authentication implementation described in DMTN-169_, Butler then has to decide whether to permit an operation on a collection following requirement DMS-REQ-0340 in LSE-61_.

Reviewer: DMTN-169 says that access to the database will go through a server that performs authorization checks and that pre-signed URLs will be returned to the user to allow authorized access to datasets.  But neither that document nor this one presents an argument that solely determining authorization by collection is sufficient to meet requirement 3 in [DMTN-169 §3.1](https://dmtn-169.lsst.io/#requirements).

Among other things, such an argument would seem to have to say that all operations require a collection and that no datasets can be added to a collection (directly or via collection chaining) unless the user has authorization to access them.

Author: I added some language saying that all Butler operations are done via a collection.

This document proposes having the Butler service maintain a list of groups and do a group membership check of the requesting user.
Other options and their trade-offs are also discussed.

.. _DMTN-169: https://dmtn-169.lsst.io/
.. _LSE-61: https://docushare.lsst.org/docushare/dsweb/Get/LSE-61

Requirements
============

DMS-REQ-0340 states:

    All Level 3 data products shall be configured to have the ability to have access restricted to the owner, a list of people, a named group, or be completely public.

In addition, some collections will be public and read-only, such as data releases.

A local Butler instance does not impose any access control and instead relies on the underlying file system permissions.
However, the Butler service as described in DMTN-169_ will receive authenticated requests from users and must decide whether to allow an operation while meeting the DMS-REQ-0340 requirement.

The Butler organizes data into collections, which support read and write operations.

> Collections are organized hierarchically has paths.

Reviewer: ```suggestion
Collections are organized hierarchically as paths.
```

> The current plan is to have the Butler service support two main path structures: one with a top-level directory per user, and one for collaborations where the top-level directory corresponds to a group.

Reviewer: ```suggestion
The current plan is to have the Butler service support two user-writable path structures in addition to common data products: one with a top-level directory per user, and one for collaborations where the top-level directory corresponds to a group.
```

A user of the Butler service will obtain a token from the identity management system and present it as part of an API call.
The Butler service can, in turn, use that token to obtain the identity of the user and their group membership using the token API (see SQR-049_).
With that information, it must somehow decide whether an operation is permitted.

.. _SQR-039: https://sqr-039.lsst.io/
.. _SQR-044: https://sqr-044.lsst.io/
.. _SQR-049: https://sqr-049.lsst.io/

Proposed solution
=================

Assume that Butler collection paths are of the form:

- ``/u/{username}/...``
- ``/g/{group}/...``
- ``/other/...``

with arbitrary structure below those top-level paths.
The exact naming may be different.
The important semantics are that there are two top-level path patterns, one for user-owned hierarchies and one for group-owned hierarchies, and then a third type of path that doesn't match either of those patterns.

This proposal assumes that paths that do not begin with ``/u`` or ``/g`` will be written outside of the Butler service (via direct access to the Butler registry, for example) and will otherwise be public and read-only.

> Each collection may optionally be associated with an :abbr:`ACL (Access Control List)`.

Reviewer: ```suggestion
Each collection may optionally be associated in the Butler registry with an :abbr:`ACL (Access Control List)`.
```

An ACL contains a list of groups that have access to that collection.

When the Butler receives a user request, it first checks whether the request is to one of the public collections.

1. If the path does not begin with ``/u`` or ``/g``, read access is allowed and write access is denied.

If no access control decision is yet made, the Butler service then uses the associated token to obtain metadata about the user by making a call to the ``/auth/api/v1/user-info`` endpoint (see SQR-049_).
This will return information such as:

.. code-block:: json

   {
     "username": "alice",
     "name": "Alice Example",
     "uid": 124187,
     "groups": [
       {
         "id": 124187,
         "name": "alice",
       },
       {
         "id": 204173,
         "name": "example-group"
       },
       {
         "id": 205671,
         "name": "other-group"
       }
     ]
   }

The Butler service may (and generally should) cache this information for a given token.

> The cache should have a maximum lifetime of an hour.

Reviewer: Maybe explain why 1 hour is appropriate and the tradeoffs of longer or shorter?

Author: I reconciled this with LSE-279 (making it 30 minutes) and added a reference.

Consider using a shorter cache timeout if the result of the authorization check is to deny access, since the most likely failure is that a person was just added to a group.

It then makes two more checks:

2. If the path starts with ``/u/{username}``, access will be granted and the collection created if the ``username`` attribute of the supplied token matches ``{username}``.
3. If the path starts with ``/g/{group}``, access will be granted if ``{group}`` appears as the ``name`` field of a group in the ``groups`` attribute of the supplied token.

> If neither of these rules grants access, the Butler will then check the ACL again:

Reviewer: ```suggestion
If neither of these rules grants access, the Butler will then check the ACL:
```

   If the intersection is non-empty, access is allowed; otherwise, access is denied.
   (GIDs should be used instead of group names, but the necessary API is not yet available in the identity management system.
   See ref:`future-work`.)

If there is no ACL, access is denied.

The Butler will also support an additional API call to set or modify the ACL for a collection.
This action is authorized using only authorization rules 2 and 3 (the rules based on the path structure).
In other words, only the user or group who owns the collection, because the collection is in their data area, can change the ACL.
Members of the ACL cannot change the ACL.
(Administrators of the Butler can of course bypass this and make ACL changes directly if necessary.)

The identity management system will guarantee that every user is also the sole member of a group whose name matches the username.
(This is desirable anyway for POSIX file system semantics for the Notebook Aspect of the Rubin Science Platform.)
Therefore, to grant a specific user access to a collection, the username of the user can be added to the ACL alongside any other group.

> This satisfies the DMS-REQ-0340 requirement.

Reviewer: ```suggestion
In addition, every user is a member of a single group representing all authorized users.
A collection can then be made public by adding this group to the ACL.
This satisfies the DMS-REQ-0340 requirement.
```

This document proposes having the Butler service maintain a list of groups and do a group membership check of the requesting user.
Other options and their trade-offs are also discussed.

.. _DMTN-169: https://dmtn-169.lsst.io/
.. _LSE-61: https://docushare.lsst.org/docushare/dsweb/Get/LSE-61

Requirements
============

DMS-REQ-0340 states:

    All Level 3 data products shall be configured to have the ability to have access restricted to the owner, a list of people, a named group, or be completely public.

Per LPM-231_, "Level 3" here refers to User Generated data products.
The Butler service will also serve Data Release data products.
Those will be public and read-only.

.. _LPM-231: https://docushare.lsst.org/docushare/dsweb/Get/LPM-231

A local Butler instance does not impose any access control and instead relies on the underlying file system permissions.
However, the Butler service as described in DMTN-169_ will receive authenticated requests from users and must decide whether to allow an operation while meeting the DMS-REQ-0340 requirement.

The Butler organizes data into collections, which support read and write operations.

> All Butler operations are done on collections.

Reviewer 2: I think we'll need to expand that; _most_ operations are, and all files and file operations are, but some of the entities that provide organizational structure for datasets (dataset types and dimension rows) are not associated with collections.

The same ownership model we use for collections will probably work for dataset types as well - those are not currently namespaced the way collections are in our internal usage, but they clearly need to be in this context.  So I think there's nothing qualitatively new about that extension.

Dimension rows will be a bit trickier, as much of our data model really depends on them being a shared resource.  They're also an "advanced usage" concept, in that most users will not ever need to define their own dimensions, and we can probably defer providing support for user-created dimensions for a while.  The main high-level use case that is difficult to support without user-owned (or at least user-inserted) dimensions is user ingest of raw data from other instruments in a way that would allow them to be processed via our middleware.  Happy to expound on the problem at some point, but a comment is probably not the most efficient method.

Reviewer 2: Actually, dataset types may be even simpler than that, because they can probably always be world-readable, so we may just be able to use the user/group namespacing rules with no ACLs.

Author: We only care about data products from a DMS-REQ-0340 perspective, so my inclination is to change this sentence to instead say "All Butler data operations are done on collections." It doesn't sound like there's much security concern about access to dataset types or dimension rows (correct me if I'm wrong), so I think you can handle them however you want.

If you want the ACL policy for them to be recorded in this document as well so that everything's in the same place, could you propose what you want to say? If I'm understanding the problem correctly, I think saying whatever you would prefer to say is fine.

Reviewer 2: Sounds good.  Might be best for @timj and I to just create a new ticket or two, for one of us to add those details to this document once we've thought them through, and ask you to make sure nothing raises any flags.

Author: Okay, I'll change this to data operations for now, and then am happy to review those additions.

Drawbacks:

- Centralized authorization systems are disfavored in security design because authorization logic is deeply tied to the data and operations model of a service.
  This is known to lead to maintainability issues.
  Either the authorization system and all the services it protects have to constantly change in lockstep, requiring coordinating changes and deployments to the authorization service when adding new features to any other service, or the authorization service has to use a complex abstract grammar of subjects, verbs, and objects in which to express generic authorization rules.
  The latter adds complexity and confusion without saving effort; defining the verbs and objects tends to be more tedious than directly implementing the authorization logic.

Pure group semantics
--------------------

Do not expose a group whose name matches the name of the user.
Instead, only support ad hoc groups.
If a user wishes to grant access to a collection to a set of people who are not already represented by a group, have them create a group and populate it with the users to whom they want to grant access.

Advantages:

> - Simplifies group semenatics.

Reviewer 3: Typo: semenatics
